components/resource_metering/src/reporter.rs

```diff
@@ -0,0 +1,211 @@
+            Task::CpuRecords(_) => {
+                write!(f, "CpuRecords")?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+pub struct ResourceMeteringReporter {
+    config: Config,
+
+    env: Arc<Environment>,
+    security_mgr: Arc<SecurityManager>,
+
+    scheduler: Scheduler<Task>,
+
+    // TODO: mock client for testing
+    client: Option<ResourceUsageAgentClient>,
+    cpu_records_collector: Option<CollectorHandle>,
+
+    // resource_tag -> ([timestamp_secs], [cpu_time_ms], total_cpu_time_ms)
+    records: HashMap<Vec<u8>, (Vec<u64>, Vec<u32>, u32)>,
+
+    find_top_k: Vec<u32>,
+}
+
+impl ResourceMeteringReporter {
+    pub fn new(
+        config: Config,
+        scheduler: Scheduler<Task>,
+        env: Arc<Environment>,
+        security_mgr: Arc<SecurityManager>,
+    ) -> Self {
+        Self {
+            config,
+            env,
+            security_mgr,
+            scheduler,
+            client: None,
+            cpu_records_collector: None,
+            records: HashMap::default(),
+            find_top_k: Vec::default(),
+        }
+    }
+
+    pub fn init_reporter(&mut self, addr: &str) {
+        let channel = {
+            let cb = ChannelBuilder::new(self.env.clone())
+                .keepalive_time(Duration::from_secs(10))
+                .keepalive_timeout(Duration::from_secs(3));
+            self.security_mgr.connect(cb, addr)
+        };
+        self.client = Some(ResourceUsageAgentClient::new(channel));
+        if self.cpu_records_collector.is_none() {
+            self.cpu_records_collector = Some(register_collector(Box::new(
+                CpuRecordsCollector::new(self.scheduler.clone()),
+            )));
+        }
+    }
+}
+
+impl Runnable for ResourceMeteringReporter {
+    type Task = Task;
+
+    fn run(&mut self, task: Self::Task) {
+        match task {
+            Task::ConfigChange(new_config) => {
+                if !new_config.should_report() {
+                    self.client.take();
+                    self.cpu_records_collector.take();
+                } else if new_config.agent_address != self.config.agent_address
+                    || new_config.enabled != self.config.enabled
+                {
+                    self.init_reporter(&new_config.agent_address);
+                }
+
+                self.config = new_config;
+            }
+            Task::CpuRecords(records) => {
+                let timestamp_secs = records.begin_unix_time_secs;
+
+                for (tag, ms) in &records.records {
+                    let tag = &tag.infos.extra_attachment;
+                    if tag.is_empty() {
+                        continue;
+                    }
+
+                    let ms = *ms as u32;
+                    match self.records.get_mut(tag) {
+                        Some((ts, cpu_time, total)) => {
+                            if *ts.last().unwrap() == timestamp_secs {
+                                *cpu_time.last_mut().unwrap() += ms;
```

Comment:
In what case will this happen?

Reply:
Records with the same `extra_attachment` but different ids in a batch